File: dsa-problems/LinkedLists/solutions/javascript/LNK-004-hostel-cleanup-deduplicate-two.js
const readline = require("readline");

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

function deduplicateAtMostTwo(head) {
  if (!head || !head.next) return head;

  let prev = head;
  let current = head.next;
  let count = 1;

  while (current !== null) {
    if (current.val === prev.val) {
      count++;
      if (count > 2) {
        // Remove current
        prev.next = current.next;
        current = current.next;
      } else {
        prev = current;
        current = current.next;
      }
    } else {
      count = 1;
      prev = current;
      current = current.next;
    }
  }
  return head;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/).filter(x => x)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  
  const dummy = new ListNode(0);
  let cur = dummy;
  for (let i = 0; i < n; i++) {
    cur.next = new ListNode(parseInt(data[idx++], 10));
    cur = cur.next;
  }

  let head = deduplicateAtMostTwo(dummy.next);
  const out = [];
  while (head) {
    out.push(head.val);
    head = head.next;
  }
  console.log(out.join(" "));
});

File: dsa-problems/LinkedLists/solutions/python/LNK-004-hostel-cleanup-deduplicate-two.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

def deduplicate_at_most_two(head: ListNode) -> ListNode:
    if not head or not head.next:
        return head
    
    prev = head
    current = head.next
    count = 1
    
    while current:
        if current.val == prev.val:
            count += 1
            if count > 2:
                # Remove current
                prev.next = current.next
                current = current.next
            else:
                # Keep current
                prev = current
                current = current.next
        else:
            # New value
            count = 1
            prev = current
            current = current.next
            
    return head

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        dummy = ListNode()
        cur = dummy
        for _ in range(n):
            cur.next = ListNode(int(next(iterator)))
            cur = cur.next
            
        head = deduplicate_at_most_two(dummy.next)
        
        out = []
        while head:
            out.append(str(head.val))
            head = head.next
        print(" ".join(out))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
